node/as/thrift.js

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var assert = require('assert');
+
+module.exports = TChannelAsThrift;
+
+function TChannelAsThrift(opts) {
+    var self = this;
+    assert(opts.spec, 'TChannelAsThrift expected spec');
+    self.spec = opts.spec;
+}
+
+TChannelAsThrift.prototype.register =
+function register(channel, name, opts, handle) {
+    var self = this;
+
+    var argsName = name + '_args';
+    var argsType = self.spec.getType(argsName);
+
+    var returnName = name + '_result';
+    var resultType = self.spec.getType(returnName);
+
+    channel.handler.register(name, handleThriftRequest);
+
+    function handleThriftRequest(req, res, inHeadBuffer, inBodyBuffer) {
+
+        // Process incoming thrift body
+        var inBodyResult = argsType.fromBuffer(inBodyBuffer);
+        if (inBodyResult.error) {
+            return inBodyResult.toCallback(handleThriftResponse);
+        }
+        var inBody = inBodyResult.value;
+
+        // TODO process inHeadBuffer into inHead
+        var inHead = null;
+
+        handle(opts, inHead, inBody, handleThriftResponse);
+
+        function handleThriftResponse(error, response) {
+            if (error) {
+                return res.sendError('UnexpectedError', error.message);
```

Comment:
:+1: BadRequest for thrift parsing; UnexpectedError for any `cb(err)` in the application developers handler function.